### DISCUSSION TITLE
Why are button colors so hard to do?

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#1920] [Q&A] [answered]
### Question

You'd think button colors would be the easiest thing in the world to set.  Then why is it so hard??

# Attempt 1
```py
ui.button ('foo').classes ('bg-red-300')
ui.run ()
# button is blue.  NO RED FOR YOU!!
```

# Attempt 2
```py
button = ui.button ('foo')
button.tailwind ('bg-red-300')
ui.run ()
# button is blue.  NO RED FOR YOU!!
```

# Attempt 3
```py
button = ui.button ('foo', color = 'bg-red-300')
ui.run ()
# button is white.  NO RED FOR YOU!!
```

# Attempt 4
Button color is background color not foreground (text) color.  We should use bg-red-300 not red-300.  Well nothing else worked, let's try it.

```py
button = ui.button ('foo', color = 'red-300')
ui.run ()
# button is red?  what the what?
```
Did we finally fix it?

# Attempt 5
Let's change the color later.
```py
button = ui.button ('foo', color = 'red-300')
button.classes ('green-300')

ui.run ()
# button is red.  NO CHANGE FOR YOU!
```

# Attempt 6
What if we do this...
```py
button = ui.button ('foo', color = 'red-300')
button.tailwind ('green-300')

ui.run ()
# button is red.  NO CHANGE FOR YOU!
```

# Question
How in god's name are you supposed to change button color after object construction?  And why are colors so hard to use??

**Comment by falkoschindler** [comment, marked as answer]
That's a really good question! And I hate that buttons are so hard to colorize.

I'm currently not at my computer for some days, so I can't respond in full length. But let me drop this link to a discussion where I explained the difficulty with button colors recently:
https://github.com/zauberzeug/nicegui/issues/1821#issuecomment-1768734775

We definitely need a better solution. I just don't know how.

**Reply by ed2050** [reply]
Thank you @falkoschindler!!  That syntax for !important is very good to know.  Oddly the colors work as intended with it: !green-300 changes button font color and !bg-green-300 changes button background color.

This should be documented on the button page if possible.  But I'm worried about you guys - do you and @rodja ever take a day off?  Hope so, you deserve it!

**Reply by ed2050** [reply]
I found a silent error:
```py
ui.button ( 'Go' , color = '!bg-red-500')
```

Not only does it not change the color, but bg-red-500 isn't anywhere in the generated html tags.  It just gets silently swallowed.

I'm guessing the color= param doesn't recognize ! like .classes() does.  That's fine.  But shoudn't color= raise some feedback?  Throw an exception, print warning to terminal, something.

**Reply by falkoschindler** [reply]
Here is the code that interprets the color values:
https://github.com/zauberzeug/nicegui/blob/6442720e9016318196aac6e50c370ae548a1f595/nicegui/elements/mixins/color_elements.py#L21-L26

As a last resort it falls back to CSS. And since "!bg-red-500" is no valid Quasar or Tailwind color, the code assumes it must be a CSS expression. But the browser seems to drop such an invalid CSS attribute. It is correctly sent to the client, but isn't contained in the DOM.

I don't know if we can recognize such a user error. Since a CSS color definition can contain a CSS color name, an rgb or rgba function call, CSS variables or who knows what else, I decided to simply swallow everything. Otherwise we would need to parse the CSS attribute to spot potential errors.
